Forge.TreeWalker.UnitTests/test/TreeSchemaValidatorTests.cs

```diff
@@ -0,0 +1,166 @@
+            schemaDirectoryPath = Path.Combine(Environment.CurrentDirectory, "test\\ExampleSchemas");
+            stringRules = File.ReadAllText(Path.Combine(Environment.CurrentDirectory, "contracts\\ForgeSchemaValidationRules.json"));
+            rulesForDictionary = File.ReadAllText(Path.Combine(Environment.CurrentDirectory, "contracts\\ForgeSchemaDictionaryValidationRules.json"));
+            jschemaRules = JSchema.Parse(stringRules);
+            forgeTreeAsString = File.ReadAllText(forgeTreeFromPath);
+            forgeTreeDictionaryAsString = File.ReadAllText(Path.Combine(Environment.CurrentDirectory, "test\\ExampleSchemas\\SubroutineSchema.json"));
+            treeSchema = JsonConvert.DeserializeObject<ForgeTree>((string)forgeTreeAsString);
+            treeSchemas = new Dictionary<string, ForgeTree>();
+            treeSchemas.Add("tree1", treeSchema);
+            treeSchemas.Add("tree2", treeSchema);
+            treeSchemas.Add("tree3", treeSchema);
+            invalidSchemaWithErrorContent = File.ReadAllText(Path.Combine(Environment.CurrentDirectory, "test\\InvalidTestSchemas\\InvalidTestSchemaErrorContent.json"));
+            invalidSchemaDirectoryPath = "test\\ExampleSchemas\\TardigradeSchema.json";
+        }
+
+        [TestMethod]
+        public void Test_GetLinkedJSchemaRules()
+        {
+            try
+            {
+                JSchema linkedRules = ForgeSchemaValidator.GetLinkedJSchemaRules(stringRules, stringRules, "//ForgeSchemaValidationRules.json");
+                ForgeSchemaValidator.ValidateSchemaAsForgeTree(treeSchema, linkedRules, out IList<ValidationError> errorList);
+            }
+            catch (Exception ex)
+            {
+                Assert.Fail("Expected no exception, but got: " + ex.Message);
+            }
+        }
+
+        [TestMethod]
+        public void Test_ValidateSchemaAsForgeTree_WithRulesAsString()
+        {
+            bool res = ForgeSchemaValidator.ValidateSchemaAsForgeTree(treeSchema, stringRules, out IList<ValidationError> errorList);
+            Assert.AreEqual(true, res);
+            Assert.AreEqual(0, errorList.Count);
+        }
+
+        [TestMethod]
+        public void Test_ValidateSchemaAsForgeTree_WithRulesAsJSchema()
+        {
+            bool res = ForgeSchemaValidator.ValidateSchemaAsForgeTree(treeSchema, jschemaRules, out IList<ValidationError> errorList);
+            Assert.AreEqual(true, res);
+            Assert.AreEqual(0, errorList.Count);
+        }
+
+        [TestMethod]
+        public void Test_ValidateSchemaAsForgeTreeDictionary_WithoutValidateAsDictionary()
+        {
+            bool res = ForgeSchemaValidator.ValidateSchemaAsForgeTreeDictionary(treeSchemas, stringRules, false, out IList<ValidationError> errorList);
+            Assert.AreEqual(true, res);
+            Assert.AreEqual(0, errorList.Count);
+        }
+
+        [TestMethod]
+        public void Test_ValidateSchemaAsForgeTreeDictionary_WithValidateAsDictionary()
```

Comment:
>Test_ValidateSchemaAsForgeTreeDictionary_WithValidateAsDictionary [](http://example.com/codeflow?start=20&length=65)

Using rulesForDictionary shouldn't work because you didn't link the rules. The use-case for dictionary-level rules was the user should call ForgeSchemaValidator.GetLinkedJSchemaRules, and use the output of that to call a validate method.

I'm not sure how the UTs are passing. We should be attempting to reference a TreeNodeDefinition of a rules file we don't have loaded.

Please debug and figure out why the UTs are passing. Once you do, create a UT for this scenario (trying to use dictionary rules without first linking), and change the code to make it fail. Hopefully we can figure out where we are allowing or skipping validation and throw an exception instead.